### DISCUSSION TITLE
Adding to StructuredAnnotations structure in OME-TIFFMetadata

### DISCUSSION THREAD

**Discussion by michael-scott-phoseon** [libvips/libvips#2457] [General]
We need to load the OME-TIFF metadata (which I am doing with NetVips), then either:

Inject the <StructuredAnnotations> section into the metadata, or... add a <TimeStampAnnotation> to an existing <StructuredAnnotations>.  This is in .NET / C#.  It seems UTTERLY IMPOSSIBLE to find code that modifies the OME-Metadata. in C++ or C#.  What am I not getting?

**Comment by jcupitt** [comment]
Hi @michael-scott-phoseon,

ome-tiff puts the metadate into the TIFF IMAGEDESCRIPTION tag as xml, so all you need to do is read and write that.

In python (for example):

```
john@banana ~/pics/ome $ python3
Python 3.9.5 (default, May 11 2021, 08:20:37) 
[GCC 10.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.new_from_file("x.tif")
>>> x.get("image-description")
'<?xml version="1.0" encoding="UTF-8"?><!-- Warning: this comment is an OME-XML metadata block, which contains crucial dimensional parameters and other important metadata. Please edit cautiously (if at all), and back up the original data before doing so. For more information, see the OME-TIFF web site: https://docs.openmicroscopy.org/latest/ome-model/ome-tiff
...
```

Then parse and modify with some xml library. To write:

```
>>> x.set("image-description", "hello")
>>> x.write_to_file("y.tif")
>>>
```

With some extra options, depending on the style of ome-tiff you want to create.

**Reply by michael-scott-phoseon** [reply]
Ah, so there is no way to MODIFY the existing OME Metadata.  I do have the OME-Metadata info, and am using the XML now.  I was just assuming that I could just MODIFY the existing metadata, not replace it.  Thanks for that!

**Reply by michael-scott-phoseon** [reply]
John,

You got me on the right track awhile ago regarding inserting comments into OME-TIFF  Metadata.  Sadly, it was in python, and I know nothing about python.  What I need is to be able to load an ome.tiff file, insert a comment, then save the file with the same name.  I am using NetVips, but that GIT page doesn't have a discussions section, which is why I am posting here (since they only have a bugs section, but this isn't a bug, but is  question.  Below is my current method in C#.  Hopefully I've documented it well enough, as I know you aren't much of a C# person.  There are 2 TODO lines at the bottom of the method.  I am already successfully loading the file, extracting the OME Metadata, adding my comment, and injecting the new OME-Metadata (with your help from a separate thread).  Could you help me with this last part.  I'd rather save the file with the same name, as I am really saving it, then keeping it open to view the images (that part already works).

        private bool AddTimeStamp(FileInfo fi)
        {
            bool success = false;

            // OME is the set of classes created from the latest OME Schema
            // Code Source: https://github.com/boonepeter/OME-Tiff-Csharp
            // To recreate classes, run "xsd.exe ome.xsd", though this version is is about 100 lines different
            OME omeMetadata = null;

            try
            {
                using (var img = NetVips.Image.Tiffload(fi.FullName))
                {
                    object objImageDescription = img.Get("image-description");
                    if (objImageDescription == null)
                        throw new Exception("Unable to obtain OME-TIFF Metadata string from file");

                    // Serialize text into OME-TIFF structure
                    XmlSerializer Serializer = new XmlSerializer((typeof(OME)));
                    using (TextReader tr = new StringReader(objImageDescription.ToString()))
                    {
                        omeMetadata = (OME)Serializer.Deserialize(tr);
                    }

                    if (omeMetadata == null)
                        throw new Exception("Unable to deserialize OME Metadata string");

                    // Convert to List
                    List<object> lstItems = new List<object>(omeMetadata.Items.ToList());

                    // Structured Annotations
                    StructuredAnnotations structuredAnnotations =
                        (StructuredAnnotations)lstItems.FirstOrDefault(
                            i => i.GetType() == typeof(StructuredAnnotations));

                    // If StructuredAnnotations wasn't in the list, add it
                    if (structuredAnnotations == null)
                        structuredAnnotations = new StructuredAnnotations();

                    // We need the Index, so we can delete / replace it later with the NEW Metadata
                    int structuralAnnotationsIdx =
                        lstItems.FindIndex(i => i.GetType() == typeof(StructuredAnnotations));

                    // Get and / or create list of TagAnnotation Annotations
                    List<Annotation> lstAnnotations = new List<Annotation>();

                    if (structuredAnnotations.Items?.Length > 0)
                        lstAnnotations = structuredAnnotations.Items.ToList();

                    // Add TimestampAnnotation
                    lstAnnotations.Add(new TimestampAnnotation()
                    {
                        Description = $"UserName: {Environment.UserName}", 
                        Value = DateTime.Now
                    });

                    // Delete old Structured Annotations. If idx was less than zero, it didn't exist
                    if (structuralAnnotationsIdx >= 0)
                        lstItems.RemoveAt(structuralAnnotationsIdx);

                    // Add then add new Structured Annotations to OME-TIFF Metadata
                    structuredAnnotations.Items = lstAnnotations.ToArray();
                    lstItems.Add((object)structuredAnnotations);

                    // Now update Metadata
                    omeMetadata.Items = lstItems.ToArray();

                    // And serialize it to XML string
                    string newMetadata = string.Empty;
                    XmlSerializer serializer = new XmlSerializer(typeof(OME));
                    using (StringWriter sw = new StringWriter())
                    {
                        serializer.Serialize(sw, omeMetadata);
                        newMetadata = sw.ToString();
                    }

                    // Now save to COPY of original image, but with new Metadata
                    using var mutated = img.Mutate(mutable =>
                    {
                        mutable.Set(GValue.GStrType, "image-description", newMetadata);
                    });

                    // TODO: New file has new Metadata, but is a 3rd the size of the original file
                    // TODO: This is where I am lost.  mutable doesn't have any of the original images
                    //mutated.WriteToFile("C:\\temp\abcd.OME.TIFF");
                    success = true;
                }
            }
            catch (Exception ex)
            {
                Log4NetHelper.WriteExceptionToLog(Log, ex, GetType(), MethodBase.GetCurrentMethod().Name);
            }

            return success;
        }

**Reply by jcupitt** [reply]
Hi again,

You won't be able to save to the same file. libvips is lazy, meaning it streams pixels from the source to the destination, and never has the whole image in memory at the same time (this is why memuse is low). If you write to the file you are reading from, you'll delete all your source data before you start.

You need to write to a new file, then after that's all finished, delete the old file and rename. Or you can write to the same filename in a  different directory, of course.

When you save the file, you need to ask libvips to regenerate the structure. Something like:

```
mutated.WriteToFile("out.tif", pyramid=True, tile=True, tile_width=512, tile_height=512, compression="jpeg")
```

But the details depend on exactly what sort of OME-TIFF you want to make. It's a complex format. You might need subifd pyramid levels, and you might want many channels (I don't remember what sort of images you are dealing with exactly).

You may need to copy the label and macro separately. Again, it depends.

**Reply by michael-scott-phoseon** [reply]
I will give it a try in a second, but in our earlier conversation, what I found was that the other images in the original ome.tiff file were NOT in the mutated file.  Will the flags in mutated.WriteToFile do that, which I suspect won't.  If not, how can I inject the other images into out.tif.  Will the subIFDs of the pyramid be in out.tif because of "pyramid=True"

**Reply by jcupitt** [reply]
It's hard to say in the abstract because OME-TIFF has several ways of attaching macros and labels. Have you shared one of these files with me? I can't remember.
